rust/cbork-utils/src/deterministic_helper.rs

```diff
@@ -0,0 +1,933 @@
+
+/// Represents a CBOR map key-value pair where the key must be deterministically encoded
+/// according to RFC 8949 Section 4.2.3.
+///
+/// This type stores the raw bytes of both key and value to enable:
+/// 1. Length-first ordering of keys (shorter keys before longer ones)
+/// 2. Lexicographic comparison of equal-length keys
+/// 3. Preservation of the original encoded form
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct MapEntry {
+    /// Raw bytes of the encoded key, used for deterministic ordering
+    pub key_bytes: Vec<u8>,
+    /// Raw bytes of the encoded value
+    pub value: Vec<u8>,
+}
+
+impl PartialOrd for MapEntry {
+    /// Compare map entries according to RFC 8949 Section 4.2.3 rules:
+    /// 1. Compare by length of encoded key
+    /// 2. If lengths equal, compare byte wise lexicographically
+    ///
+    /// Returns Some(ordering) since comparison is always defined for these types
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for MapEntry {
+    /// Compare map entries according to RFC 8949 Section 4.2.3 rules:
+    /// 1. Compare by length of encoded key
+    /// 2. If lengths equal, compare byte wise lexicographically
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.key_bytes
+            .len()
+            .cmp(&other.key_bytes.len())
+            .then_with(|| self.key_bytes.cmp(&other.key_bytes))
+    }
+}
+
+/// Decodes a CBOR map with deterministic encoding validation (RFC 8949 Section 4.2.3)
+/// Returns the raw bytes of the map if it passes all deterministic validation rules.
+///
+/// From RFC 8949 Section 4.2.3:
+/// "The keys in every map must be sorted in the following order:
+///  1. If two keys have different lengths, the shorter one sorts earlier;
+///  2. If two keys have the same length, the one with the lower value in (byte-wise)
+///     lexical order sorts earlier."
+///
+/// Additionally:
+/// - Map lengths must use minimal encoding (Section 4.2.1)
+/// - Indefinite-length maps are not allowed (Section 4.2.2)
+/// - No two keys may be equal (Section 4.2.3)
+/// - The keys themselves must be deterministically encoded
+///
+/// # Errors
+///
+/// Returns `DeterministicError` if:
+/// - Input is empty (`UnexpectedEof`)
+/// - Map uses indefinite-length encoding (`IndefiniteLength`)
+/// - Map length is not encoded minimally (`NonMinimalInt`)
+/// - Map keys are not properly sorted (`UnorderedMapKeys`)
+/// - Duplicate keys are found (`DuplicateMapKey`)
+/// - Map key or value decoding fails (`DecoderError`)
+pub fn decode_map_deterministically(d: &mut Decoder) -> Result<Vec<u8>, minicbor::decode::Error> {
+    validate_input_not_empty(d)?;
+
+    // From RFC 8949 Section 4.2.2:
+    // "Indefinite-length items must be made definite-length items."
+    // The specification explicitly prohibits indefinite-length items in
+    // deterministic encoding to ensure consistent representation.
+    match d.datatype()? {
```

Comment:
I think this particular check is redundant and could be done during the `d.map()` method execution.
Something like this.
```rust
let map_len = d.map()?.ok_or(minicbor::decode::Error::message(
                "Indefinite-length items must be made definite-length items",
            ))?;
```

Reply:
resolved